runtime/jcl/common/sun_misc_Unsafe.cpp

```diff
@@ -847,6 +847,85 @@ Java_jdk_internal_misc_Unsafe_objectFieldOffset1(JNIEnv *env, jobject receiver,
 	return offset;
 }
 
+/**
+ * Writes modified memory in an address range from cache to main memory.
+ * Uses memory barriers before and after writeback to ensure ordering.
+ * 
+ * On x86, the writeback is done using the CLWB instruction if available
+ * for performance, falling back to CLFLUSHOPT then CLFLUSH otherwise.
+ * 
+ * @param addr address of block to write back to memory
+ * @param len length of block being written
+ */
+void JNICALL
+Java_jdk_internal_misc_Unsafe_writebackMemory(JNIEnv *env, jobject receiver, jlong addr, jlong len)
+{
+#if defined(J9X86) || defined(J9HAMMER)
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9JavaVM *vm = currentThread->javaVM;
+
+	if (vm->dCacheLineSize > 0) {
+		uintptr_t cache_line_size = vm->dCacheLineSize;
+		U_8 *end_addr = (*(U_8 **) &addr) + (*(uintptr_t *) &len) - 1;
+		U_8 *ptr = (U_8 *)((*(uintptr_t *) &addr) & ~(cache_line_size - 1));
+		if (vm->cpuCacheWritebackCapabilities == J9PORT_X86_FEATURE_CLWB) {
+			issueReadWriteBarrier();
+			for (; ptr < end_addr; ptr += cache_line_size) {
+				asm volatile("clwb %0" : "+m" (*ptr));
+			}
+			/* writeback any partial cache line at the end */
+			asm volatile("clwb %0" : "+m" (*end_addr));
+			issueReadWriteBarrier();
+			return;
+		} else if (vm->cpuCacheWritebackCapabilities == J9PORT_X86_FEATURE_CLFLUSHOPT) {
+			issueReadWriteBarrier();
+			for (; ptr < end_addr; ptr += cache_line_size) {
+				asm volatile("clflushopt %0" : "+m" (*ptr));
+			}
+			asm volatile("clflushopt %0" : "+m" (*end_addr));
+			issueReadWriteBarrier();
+			return;
+		} else if (vm->cpuCacheWritebackCapabilities == J9PORT_X86_FEATURE_CLFSH) {
+			issueReadWriteBarrier();
+			for (; ptr < end_addr; ptr += cache_line_size) {
+				asm volatile("clflush %0" : "+m" (*ptr));
+			}
+			asm volatile("clflush %0" : "+m" (*end_addr));
+			issueReadWriteBarrier();
+			return;
+		}
+	}
+#endif /* x86 */
+
+	jclass exceptionClass = env->FindClass("java/lang/RuntimeException");
+	if (exceptionClass == NULL) {
+		/* Just return if we can't load the exception class. */
+		return;
+	}
+	env->ThrowNew(exceptionClass, "writebackMemory not supported");
+}
+
+/**
+ * Checks if the platform supports writeback from cache to memory. This
+ * function checks if the necessary CPU or OS features for writeback are
+ * enabled.
+ * On x86, these are the CLFLUSH, CLFLUSHOPT and CLWB instructions.
+ */
+jboolean JNICALL
+Java_jdk_internal_misc_Unsafe_isWritebackEnabled(JNIEnv *env, jclass clazz)
+{
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	jboolean result = JNI_FALSE;
+#if defined(J9X86) || defined(J9HAMMER)
+	result = vm->dCacheLineSize > 0 &&
```

Comment:
Please bracket all of the subexpressions, or better yet, make this an if followed by a switch, with a single TRUE assignment inside.
```
if (vm->dCacheLineSize > 0) {
   switch(vm->cpuCacheWritebackCapabilities) {
      case J9PORT_X86_FEATURE_CLWB:
      case J9PORT_X86_FEATURE_CLFLUSHOPT:
      case J9PORT_X86_FEATURE_CLFSH:
         result = JNI_TRUE;
         break;
      default:
         break;
   }
}
```

Reply:
Changed to use a switch.